src/cmap/connection_pool.ts

```diff
@@ -572,6 +594,50 @@ export class ConnectionPool extends TypedEventEmitter<ConnectionPoolEvents> {
     });
   }
 
+  /**
+   * Reauthenticate on the same connection and then retry the operation.
+   */
+  private reauthenticate(
+    connection: Connection,
+    fn: WithConnectionCallback,
+    callback: Callback
+  ): void {
+    const authContext = connection.authContext;
+    if (!authContext) {
+      return callback(new MongoRuntimeError('No auth context found on connection.'));
+    }
+    authContext.reauthenticating = true;
```

Comment:
just a thought to avoid leaking details about how the auth context manages state when re-authenticating.  what if we added a `reauth` method the `AuthProvider`?

```typescript
class MongoDBOIDC {
  reauth(authContext, callback) {
    authContext.reauthenticating = true;
    const _cb = (error, result) => {
      authContext.reauthenticating = false;
      _cb(error, result)
    }
    authContext.execute(authContext, _cb)
  }
}
```

another advantage of this approach is that it immediately becomes clear which auth mechanisms support re-authentication, because `reauth` could throw an error if we attempt to reauth using a mechanism that doesn't support it

Reply:
Added `reauth` to the `AuthProvider` but not throwing an error since all mechanisms can reauth - there's just a slight difference in behaviour when using OIDC with the cache. All other mechanisms just start auth over from the beginning.